ddtrace/internal/_file_queue.py

```diff
@@ -0,0 +1,63 @@
+import os
+import os.path
+import tempfile
+import typing
+
+from ddtrace.internal._unpatched import unpatched_open
+
+
+try:
+    # Unix based file locking
+    # Availability: Unix, not Emscripten, not WASI.
+    import fcntl
+
+    def lock(f):
+        fcntl.lockf(f, fcntl.LOCK_EX)
+
+    def unlock(f):
+        fcntl.lockf(f, fcntl.LOCK_UN)
+
+except ModuleNotFoundError:
+    # Availability: Windows
+    import msvcrt
+
+    def size(f):
+        return os.path.getsize(os.path.realpath(f.name))
+
+    def lock(f):
+        msvcrt.locking(f.fileno(), msvcrt.LK_RLCK, size(f))
+
+    def unlock(f):
+        msvcrt.locking(f.fileno(), msvcrt.LK_UNLCK, size(f))
+
+
+class File_Queue:
+    """A simple file-based queue implementation for multiprocess communication."""
+
+    def __init__(self) -> None:
+        self.directory = tempfile.gettempdir()
+        self.filename = os.path.join(self.directory, "file_queue")
+
+    def put(self, data: str) -> None:
+        """Push a string to the queue."""
+        try:
+            with unpatched_open(self.filename, "a") as f:
+                lock(f)
+                f.write(data + "\x00")
+                unlock(f)
+        except Exception:  # nosec
+            pass
+
+    def get_all(self) -> typing.Set[str]:
+        """Pop all unique strings from the queue."""
+        try:
+            with unpatched_open(self.filename, "r+") as f:
+                lock(f)
```

Comment:
Do we need to lock to read? I wonder what happens if there are many services and a lot of contention here

Reply:
this is the only way to not miss a service name, as the read also delete the file to clear the queue

Reply:
Please note that the read should not happen more than every 5sec, as it's only done by the main thread when sending data through RC.